src/recommendation/src/main/scala/RankingAdapter.scala

```diff
@@ -0,0 +1,382 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in project root for information.
+
+package com.microsoft.ml.spark
+
+import org.apache.spark.ml.param._
+import org.apache.spark.ml.recommendation._
+import org.apache.spark.ml.util._
+import org.apache.spark.ml.{Estimator, Model, Transformer}
+import org.apache.spark.sql._
+import org.apache.spark.sql.expressions.Window
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.functions.{collect_list, rank => r}
+import org.apache.spark.sql.types.{ArrayType, FloatType, IntegerType, StructType}
+
+import scala.collection.mutable
+import scala.util.Random
+
+trait RankingParams extends Params {
+  val minRatingsPerUser: IntParam = new IntParam(this,
+    "minRatingsPerUser",
+    "min ratings for users > 0",
+    ParamValidators.inRange(0, Integer.MAX_VALUE))
+
+  val minRatingsPerItem: IntParam = new IntParam(this,
+    "minRatingsPerItem",
+    "min ratings for items > 0",
+    ParamValidators.inRange(0, Integer.MAX_VALUE))
+
+  /** @group setParam */
+  def setMinRatingsPerUser(value: Int): this.type = set(minRatingsPerUser, value)
+
+  /** @group setParam */
+  def setMinRatingsPerItem(value: Int): this.type = set(minRatingsPerItem, value)
+
+  /** @group getParam */
+  def getMinRatingsPerUser: Int = $(minRatingsPerUser)
+
+  /** @group getParam */
+  def getMinRatingsPerItem: Int = $(minRatingsPerItem)
+
+  val recommender = new EstimatorParam(this, "recommender", "estimator for selection", { x: Estimator[_] => true })
+
+  /** @group getParam */
+  def getRecommender: Estimator[_ <: Model[_]] = $(recommender)
+
+  /** @group setParam */
+  def setRecommender(value: Estimator[_ <: Model[_]]): this.type = set(recommender, value)
+
+  val k: IntParam = new IntParam(this, "k", "number of items to recommend")
+
+  /** @group getParam */
+  def getK: Int = $(k)
+
+  /** @group setParam */
+  def setK(value: Int): this.type = set(k, value)
+
+}
+
+trait RankingFunctions extends RankingParams with HasRecommenderCols {
+
+  private def filterByItemCount(dataset: Dataset[_], itemCol: String, userCol: String): DataFrame =
+    dataset
+      .groupBy(userCol)
+      .agg(col(userCol), count(col(itemCol)).alias("nitems"))
+      .where(col("nitems") >= $(minRatingsPerUser))
```

Comment:
would be nice to move these to some constants (this is scala and not python so there will be compile time errors instead of runtime so it makes sense to have constants)

Reply:
I don't understand. Should $(minRatingsPerUser) be a val? or should nitems be a constant?